### PR TITLE
Hw 45 spring boot shopping cart model

### PR DESCRIPTION
**Pull request by GoldHolub**
Added ShoppingCart model to book-store-app

### CODE REVIEW

File: src/main/java/com/example/demo/service/impl/ShoppingCartServiceImpl.java [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+    }
+
+    @Override
+    public void deleteBookFromTheShoppingCart(Authentication authentication, Long cartItemId) {
+        User user = (User) authentication.getPrincipal();
+        CartItem cartItem = cartItemRepository.findById(cartItemId).orElseThrow(
+                () -> new EntityNotFoundException("Can't find CartItem by id: " + cartItemId));
+        ShoppingCart shoppingCart = shoppingCartRepository.findByUserId(user.getId()).orElseThrow(
+                () -> new EntityNotFoundException("Can't find ShoppingCart by userId: "
+                        + user.getId()));
```
Comment (okuzan) [inline src/main/java/com/example/demo/service/impl/ShoppingCartServiceImpl.java]: extract repeated logic to private utility methods

File: src/main/resources/db/changelog/changes/07-create-ShoppingCart-and-CartItem-tables.yaml [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+              constraints:
+                foreignKeyName: fk_cart_items_book_id
+                references: books(id)
+                onDelete: CASCADE
+                nullable: false
+          - column:
+              name: quantity
+              type: INT
+              constraints:
+                nullable: false
```
Comment (okuzan) [inline src/main/resources/db/changelog/changes/07-create-ShoppingCart-and-CartItem-tables.yaml]: make sure to leave exactly one empty line at the EOF

File: src/main/resources/db/changelog/db.changelog-master.yaml [hunk 3]
```diff
@@ -11,3 +11,5 @@ databaseChangeLog:
       file: db/changelog/changes/05-add-roles-and-users.yaml
   - include:
       file: db/changelog/changes/06-create-categories-and-books_categories-tables.yaml
+  - include:
+      file: db/changelog/changes/07-create-ShoppingCart-and-CartItem-tables.yaml
```
Comment (okuzan) [inline src/main/resources/db/changelog/db.changelog-master.yaml]: same

**Review by okuzan — CHANGES_REQUESTED** [review]

File: src/main/java/com/example/demo/controller/ShoppingCardController.java [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+                                @RequestBody @Valid CartItemRequestDto requestDto) {
+        return shoppingCartService.addBookToShoppingCart(authentication, requestDto);
+    }
+
+    @PutMapping("/cart-items/{cartItemId}")
+    @Operation(summary = "Update a book in ShoppingCart",
+            description = "update a book in ShoppingCart")
+    public CartItemResponseDto updateBookQuantityInShoppingCart(Authentication authentication,
+                                           @PathVariable Long cartItemId,
+                                           @RequestBody CartItemUpdateDataRequestDto requestDto) {
```
Comment (sokimaaa) [inline src/main/java/com/example/demo/controller/ShoppingCardController.java]: Validation here?

File: src/main/java/com/example/demo/dto/shoppingcard/cartitem/CartItemRequestDto.java [hunk 5]
```diff
@@ -0,0 +1,15 @@
+package com.example.demo.dto.shoppingcard.cartitem;
+
+import jakarta.validation.constraints.Min;
+import jakarta.validation.constraints.NotNull;
+import lombok.Data;
+
+@Data
```
Comment (sokimaaa) [inline src/main/java/com/example/demo/dto/shoppingcard/cartitem/CartItemRequestDto.java:7]: consider using only one approach throughout entire project:
1. creating dto as POJO
2. creating dto as `record`

File: src/main/java/com/example/demo/model/Book.java [hunk 6]
```diff
@@ -38,11 +39,11 @@ public class Book {
     private String coverImage;
     @Column(name = "is_deleted", nullable = false)
     private boolean isDeleted = false;
-    @ManyToMany
     @JoinTable(
             name = "books_categories",
             joinColumns = @JoinColumn(name = "book_id"),
             inverseJoinColumns = @JoinColumn(name = "category_id")
     )
+    @ManyToMany(fetch = FetchType.EAGER)
```
Comment (sokimaaa) [inline src/main/java/com/example/demo/model/Book.java]: use Lazy fetch type for learning purposes

File: src/main/java/com/example/demo/service/impl/ShoppingCartServiceImpl.java [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+        User user = (User) authentication.getPrincipal();
+        CartItem cartItem = getValidCartItem(cartItemId);
+        ShoppingCart shoppingCart = getValidShoppingCart(user.getId());
+        if (cartItemRepository.findByShoppingCartId(shoppingCart.getId()).contains(cartItem)) {
+            cartItem.setQuantity(requestDto.getQuantity());
+            cartItemRepository.save(cartItem);
+        } else {
+            throw new EntityNotFoundException("Can't find CartItem by id: " + cartItemId
+                    + " in the ShoppingCart to update");
+        }
```
Comment (okuzan) [inline src/main/java/com/example/demo/service/impl/ShoppingCartServiceImpl.java]: it's better to throw exception early (invert condition) and unwrap the rest of your code to make it more readable

File: src/main/resources/db/changelog/db.changelog-master.yaml (hunk 3 above)
Comment (okuzan) [inline src/main/resources/db/changelog/db.changelog-master.yaml]: naming is a bit inconsistent, let's use lowercase with underscore, as in your previous files

File: src/main/java/com/example/demo/service/impl/ShoppingCartServiceImpl.java [hunk 8]
```diff
@@ -0,0 +1,121 @@
...
+    public void deleteBookFromTheShoppingCart(Authentication authentication, Long cartItemId) {
+        User user = (User) authentication.getPrincipal();
+        CartItem cartItem = getValidCartItem(cartItemId);
+        ShoppingCart shoppingCart = getValidShoppingCart(user.getId());
+        if (cartItemRepository.findByShoppingCartId(shoppingCart.getId()).contains(cartItem)) {
+            cartItemRepository.delete(cartItem);
+        } else {
+            throw new EntityNotFoundException("Can't find CartItem by id: " + cartItemId
+                    + " in the ShoppingCart to delete");
+        }
```
Comment (okuzan) [inline src/main/java/com/example/demo/service/impl/ShoppingCartServiceImpl.java]: same

**Review by okuzan — CHANGES_REQUESTED** [review]

**Pull request merged**